### PR TITLE
Merge #55711

### LINKED ISSUE #55655: Unexpected exception Comparing aggregate functions with different types: AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64) and AggregateFunction(stochasticLinearRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64)
**Describe what's wrong**

https://fiddle.clickhouse.com/a7468d06-8a1b-485d-8f26-1d0ed75c6d27

The query (simplified from a production issue)
```
select *
from (
		select (
				select stochasticLogisticRegressionState(0.1, 0., 5, 'SGD')(number, number)
				from numbers(10)
			) as col1,
			(
				select stochasticLinearRegressionState(0.1, 0., 5, 'SGD')(number, number)
				from numbers(10)
			) as col2
		from numbers(1)
	);       
```
throws an exception,
```
Code: 43. DB::Exception: Comparing aggregate functions with different types: AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64) and AggregateFunction(stochasticLinearRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64): While processing identity(_CAST('\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', 'AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, \'SGD\'), UInt64, UInt64)')) AS col1, identity(_CAST('\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', 'AggregateFunction(stochasticLinearRegression(0.1, 0., 5, \'SGD\'), UInt64, UInt64)')) AS col2. (ILLEGAL_TYPE_OF_ARGUMENT) (version 23.2.4.1.2310B907)
, server ClickHouseNode
```

**Does it reproduce on recent release?**
Yes.

**How to reproduce**
https://fiddle.clickhouse.com/a7468d06-8a1b-485d-8f26-1d0ed75c6d27

* Which ClickHouse server version to use
Latest

**Expected behavior**
no exception

**Error message and/or stacktrace**
```
Code: 43. DB::Exception: Comparing aggregate functions with different types: AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64) and AggregateFunction(stochasticLinearRegression(0.1, 0., 5, 'SGD'), UInt64, UInt64): While processing identity(_CAST('\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', 'AggregateFunction(stochasticLogisticRegression(0.1, 0., 5, \'SGD\'), UInt64, UInt64)')) AS col1, identity(_CAST('\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0', 'AggregateFunction(stochasticLinearRegression(0.1, 0., 5, \'SGD\'), UInt64, UInt64)')) AS col2. (ILLEGAL_TYPE_OF_ARGUMENT) (version 23.2.4.1.2310B907)
, server ClickHouseNode
```

image

**Additional context**

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Scalar subqueries returning different types of aggregate function states with the same serialized state will be correctly distinguished. This closes #55655. This closes #55711.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 3aeb155f996ec6a564074a0135ec22da526a1586 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60136/3aeb155f996ec6a564074a0135ec22da526a1586/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60136/3aeb155f996ec6a564074a0135ec22da526a1586/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by Alex-Cheng** [comment]
why nobody review the PR?

File: src/Core/Settings.h [hunk 1]
```diff
@@ -710,7 +710,7 @@ class IColumn;
     M(Bool, optimize_injective_functions_in_group_by, true, "Replaces injective functions by it's arguments in GROUP BY section", 0) \
     M(Bool, optimize_group_by_function_keys, true, "Eliminates functions of other keys in GROUP BY section", 0) \
     M(Bool, optimize_group_by_constant_keys, true, "Optimize GROUP BY when all keys in block are constant", 0) \
-    M(Bool, legacy_column_name_of_tuple_literal, false, "List all names of element of large tuple literals in their column names instead of hash. This settings exists only for compatibility reasons. It makes sense to set to 'true', while doing rolling update of cluster from version lower than 21.7 to higher.", 0) \
```
Comment (Alex-Cheng) [inline src/Core/Settings.h:713]: I am not sure whether we can change the setting without worrying about compatibility.

File: src/Core/SettingsChangesHistory.h [hunk 2]
```diff
@@ -189,7 +189,7 @@ static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> sett
     {"21.12", {{"stream_like_engine_allow_direct_select", true, false, "Do not allow direct select for Kafka/RabbitMQ/FileLog by default"}}},
     {"21.9", {{"output_format_decimal_trailing_zeros", true, false, "Do not output trailing zeros in text representation of Decimal types by default for better looking output"},
               {"use_hedged_requests", false, true, "Enable Hedged Requests feature by default"}}},
-    {"21.7", {{"legacy_column_name_of_tuple_literal", true, false, "Add this setting only for compatibility reasons. It makes sense to set to 'true', while doing rolling update of cluster from version lower than 21.7 to higher"}}},
```
Comment (Alex-Cheng) [inline src/Core/SettingsChangesHistory.h:192]: should we change the the history?

**Comment by alexey-milovidov** [comment]
Let's check if it is fixed after enabling Analyzer. Then we can avoid these changes.

**Comment by alexey-milovidov** [comment]
Yes, it is fixed by Analyzer.

**Pull request closed without merging**